### ISSUE TITLE
audit(clapkit): formalize int8 artifact gates and tower-specific evaluation

### ISSUE THREAD

**Issue by SRjoeee** [findit-studio/coremlit#30]
## Audit scope

Audited current `main` at `a9439046bf3d5ed6b5874a3e053e37eae2358c34` on an Apple M1 / 8 GB / macOS 26.5.2 host.

The audit used the **FinDIT-Studio-owned CoreML artifacts**, pinned to:

- `FinDIT-Studio/clapkit-coreml@02a99c6a8be21da1e9a947499ea503a10c80c4f1`
- source lineage: `laion/clap-htsat-unfused@8fa0f1c6d0433df6e97c127f64b2a1d6c0dcda8a`
- ONNX controls: `Xenova/clap-htsat-unfused@c28f2883575e590e04d3146ff0713c2448d691ba`

All 20 downloaded CoreML `.mlmodelc` files matched the current HF `CHECKSUMS.sha256`. All four ONNX controls matched their repository LFS SHA-256 and exact size.

## Executive verdict

**No High/Critical production correctness defect was found.** The fp16 conversion is faithful and all committed gates pass. The newly published int8 tier is usable for experimentation and preserves the tested zero-shot ranking, but it is not yet represented by a first-class artifact tier or tier-specific committed gates.

### Evidence

The code/test documentation still pins the earlier fp16-only HF revision:

- `crates/clapkit/README.md:145-188`
- `crates/clapkit/tests/common/mod.rs:3-6`
- `crates/clapkit/tests/model_io.rs:8-12`
- `crates/clapkit/tests/text_model_io.rs:6-8`

Current HF revision `02a99c6…` adds:

- `clap_audio_int8.mlmodelc`
- `clap_text_int8.mlmodelc`

The existing loader accepts them because their I/O contracts match. Real execution results:

- int8 I/O contract: PASS for both towers
- int8 placement: audio worst cross-unit cosine `0.99998385`; text `0.99993682`
- 192.5 s e2e speech: PASS; top label unchanged (`person speaking`, logit `8.885357`)
- int8 vs fp16, 3 real audio windows: worst cosine `0.99991471`
- int8 vs fp16, 12 multilingual prompts: worst cosine `0.99929160`

Reusing the **fp16-specific** parity suite against int8 yielded 4 PASS / 1 FAIL:

```text
text fp32-control worst cosine 0.99924403
existing fp16 lower bound      0.99980000
```

This does **not** prove the int8 model is broken; it proves the new tier needs its own measured bands.

### Requested closure

- Add a first-class tier/resolver such as `ModelTier::{Fp16, Int8}` or equivalent documented artifact selection.
- Add int8-specific SHA, I/O, placement, parity, e2e, and zero-shot ranking gates.
- Decide explicitly whether fp16 or int8 is the production default per tower.

### Evidence

The documented command at `crates/clapkit/README.md:185-188` downloads mutable HF `main` without `--revision`, while nearby provenance text declares revision `97d631f3…`.

Committed SHA gates cover only:

- fp16 `model.mil`
- fp16 `weights/weight.bin`

See:

- `crates/clapkit/tests/model_io.rs:63-84`
- `crates/clapkit/tests/text_model_io.rs:47-65`

They do not cover fp16 metadata/coremldata/analytics or any int8 file. This audit separately checked all 20 current `.mlmodelc` files and they passed, but CI does not currently enforce that complete manifest.

### Requested closure

- Add `--revision <immutable SHA>` to the documented download command.
- Verify every downloaded runtime file against a committed/revision-bound manifest.
- Update provenance references to the revision that includes the int8 artifacts.

### Evidence

The committed model parity corpus is narrow:

- one JFK fixture split into three windows: `tests/parity_textclap.rs:64-87`
- five text prompts: `tests/parity_textclap.rs:76-83`
- e2e creates 192.5 s by tiling the same JFK clip: `tests/e2e.rs:7-10,46-53`

Measured fidelity is excellent:

| Comparison | Audio worst cosine | Text worst cosine |
|---|---:|---:|
| fp16 CoreML vs fp32 ONNX | 0.99999756 | 0.99995059 |
| fp16 CoreML vs original quantized textclap ONNX | 0.99804741 | 0.96717650 |
| int8 CoreML vs fp32 ONNX | 0.99991775 | 0.99924403 |
| int8 CoreML vs original quantized textclap ONNX | 0.99801934 | 0.96927553 |

This proves conversion fidelity, but not broad task quality.

### Requested closure

- ESC-50 official split zero-shot accuracy + per-class confusion.
- FindIT real-audio retrieval Recall@K / mAP.
- Mixed-event long-audio evaluation of `MeanRenormalized`, `CoverageWeightedMean`, and EMA.
- Include multilingual/CJK prompt sensitivity, where the original quantized ONNX showed the largest drift.

## P1 — Make performance claims tower-specific

A local directional first pass used debug build, CoreML `ComputeUnits::All`, ONNX Runtime default CPU, one warmup, seven warm samples. These are engineering signals, not release marketing numbers.

| Tower/backend | Warm median |
|---|---:|
| CoreML audio fp16 | 855.72 ms |
| ONNX audio quantized | 934.16 ms |
| ONNX audio fp32 | 956.75 ms |
| CoreML text fp16 | 27.63 ms |
| ONNX text quantized | 4.57 ms |
| ONNX text fp32 | 9.73 ms |

Implications:

- audio CoreML was `8.40%` lower latency than quantized ONNX (`1.09×` speedup)
- text CoreML was `6.04×` slower than quantized ONNX and `2.84×` slower than fp32 ONNX
- CoreML int8 audio showed no speed gain vs fp16 in a separate run
- CoreML int8 text was about `20.97%` slower than fp16 in that separate run

Load costs were also much larger for CoreML in this process (`24.2 s` audio, `7.25 s` text) than ONNX (`0.39–0.86 s`). Production must reuse/prewarm encoders and record cache state.

Requested follow-up: add a release benchmark that separates first-ever specialization, cached load, first inference, and warm inference (>=20 runs), with output hashes/cosines and RSS.

## P2 — Known operational/legal follow-ups

1. The audio tower repeatedly emitted `ANECCompile() FAILED` and fell back to GPU/CPU. This is already honestly documented (`tests/placement.rs:8-12`) and numerical agreement remained good, but it limits ANE/efficiency claims.
2. Weight license metadata remains ambiguous: clapkit treats weights as CC-BY-4.0 while the current upstream HF card declares Apache-2.0. Attribution is present in `NOTICE`; owner/legal should reconcile the metadata before redistribution.

## Commands executed

```text
cargo test -p clapkit --all-targets
cargo clippy -p clapkit --all-targets --all-features -- -D warnings
cargo fmt --all -- --check
CLAPKIT_TEST_MODELS=<fp16> cargo test -p clapkit -- --ignored --nocapture
CLAPKIT_TEST_MODELS=<fp16> CLAPKIT_TEXTCLAP_ONNX=<onnx> cargo test -p clapkit --features parity-oracle --test parity_textclap -- --ignored --nocapture
# The model I/O, placement, e2e, and parity commands were repeated against an int8 alias directory.
```

Static result: 72/72 default tests passed; clippy/fmt/diff checks passed; final repository working tree clean.

**Comment by SRjoeee** [comment]
## Verified adversarial-review addendum

A separate read-only reviewer completed after the initial audit. I independently re-ran/checked its findings against `main@a9439046`; the following should be added to this issue.

### P0 — Make parity reducers fail closed on non-finite values

`crates/clapkit/tests/parity_textclap.rs:59-62` computes cosine without an explicit finite check, then `:100-107` and `:117-124` reduce with:

```rust
worst = worst.min(cos);
```

Rust `f32::min` returns the finite operand when the other operand is NaN. Starting at `worst = 1.0`, a NaN sample therefore leaves `worst == 1.0`, and the fp32 band `[0.9998, 1.0001]` passes.

Independent reproduction:

```text
worst_after_nan = 1.0
fp32_band_passes = true
```

The current real encoder paths normally prevent this because both clapkit and the currently configured textclap projection paths normalize through finite-checking constructors. This is therefore a **latent test false-green**, not evidence that the measured runs contained NaN. It still needs a fail-closed guard so a future oracle/runtime change cannot silently green.

Requested fix:

- assert equal lengths;
- assert every input element finite;
- assert final cosine finite;
- add a hermetic reducer test injecting NaN and requiring failure.

### P0 — Run clapkit model gates in CI

The normal job runs `cargo test --workspace`, under which the clapkit model tests remain ignored and `parity_textclap` is not built without `parity-oracle`.

The `model-tests` job in `.github/workflows/ci.yml:40-134` downloads only WhisperKit models/tokenizer and runs coremlit/whisperkit model tests. It never downloads or executes clapkit CoreML/ONNX gates.

Consequently, deleting/weaking a clapkit model test or drifting a published artifact does not currently block merge.

Requested fix: add a revision/manifest-keyed clapkit CI model job running I/O, checksum, placement-agreement, e2e, and fp16/int8 parity targets. Missing models must fail, not skip.

### P1 — Weaken the current “quantization contribution” causal language

`tests/parity_textclap.rs:17-28,160-167,189-193` labels the difference between:

- cosine(CoreML fp16, Xenova fp32), and
- cosine(CoreML fp16, Xenova quantized)

as the Xenova “quantization contribution.” This is useful evidence but not a strict causal decomposition because both values use CoreML fp16—not upstream PyTorch fp32—as the common anchor.

For strict attribution, freeze identical graph inputs and compare five arms against the same PyTorch fp32 oracle:

1. LAION PyTorch fp32
2. Xenova ONNX fp32
3. Xenova ONNX quantized
4. CoreML fp16
5. CoreML int8

Report cosine, L2, max-abs, mean-abs, norm, and finite for every arm. Until then, phrase the result as “difference associated with the Xenova quantized export relative to its fp32 sibling,” not an exhaustive causal decomposition.

### P1 — Separate numerical agreement from observed hardware placement

`tests/placement.rs` proves that outputs requested under different `ComputeUnits` agree numerically. It does not prove residency: `CpuAndNeuralEngine` can silently fall back to CPU/GPU and still pass cosine.

Rename/describe this as requested-compute-unit numerical agreement. Collect actual observed placement separately with Core ML Instruments/performance reports and retain fallback diagnostics.

### P1 — Make the conversion recipe reproducible and safe

Verified issues:

- `conversion/run_clapkit.sh:7-10` hard-codes a one-off `/private/tmp/claude-501/...` environment.
- The conversion directory has no executable dependency lock; tool versions are prose only.
- The committed recipe generates/verifies fp16/fp32 only; it does not contain the published int8 palettization recipe.
- `_clap_common.py:3-17` deliberately monkey-patches Transformers' `torch.load` CVE guard for a pickle checkpoint.
- The upstream checkpoint revision is pinned, but the conversion inputs do not independently pin the `pytorch_model.bin` LFS digest in a local manifest.

Verified upstream artifact:

```text
laion/clap-htsat-unfused@8fa0f1c6…/pytorch_model.bin
size   614,525,833 bytes
sha256 1cd3c601bc4afe0fa87be3de4c13dd2cfadd249fac1e29acf74a9b296c3219bb
```

Requested fix:

- environment/CLI-derived paths;
- `pyproject.toml` + lockfile or equivalent;
- immutable input manifest including checkpoint LFS SHA-256;
- isolated conversion environment for pickle loading;
- committed int8 palettization method/config and verifier;
- fail-closed probe thresholds with machine-readable outputs.

### Correction to the independent reviewer log

The reviewer reported a transient default doctest failure. I re-ran the canonical command on the unchanged tree:

```text
cargo test -p clapkit
71 unit tests passed
1 tokenizer integration test passed
3 doctests passed
exit code 0
```

Therefore **the current default crate test is green**. The real coverage issue is that the model-gated tests are ignored/not run in CI, not a persistent doctest failure.

**Comment by uqio** [comment]
Fully remediated by #35 (merged to main as ee50e67). All four findings addressed: fail-closed parity cosine (length+finite guard), int8 artifact gates, exact-enumerate SHA manifest (assert_exact_sha_manifest, mutation-proven), and the destructive cold-cache bench mode removed. Closing as solved.
